waiter/src/waiter/process_request.clj

```diff
@@ -590,17 +619,37 @@
                               (async/close! request-state-chan)
-                              ; Handle error lower down
-                              (throw e))))))
+                              (let [response (-> (process-exception-fn track-process-error-metrics request descriptor e)
+                                                 (update :headers (fn [headers]
+                                                                    (merge @response-headers headers))))
+                                    request (-> request
+                                                (utils/mark-request-time :closed))]
+                                (rlog/log (assoc (rlog/request->context request)
+                                                 :status (:status response)
+                                                 :termination-state :backend-error))
+                                response))))))
                     (catch Exception e
-                      ; Handle error lower down
-                      (throw (ex-info "Rethrowing exception from process" {:descriptor descriptor :source :pr/process} e)))))))
+                      (let [{:keys [descriptor]} (ex-data e)
+                            response (-> (process-exception-fn track-process-error-metrics request descriptor e)
+                                         (update :headers (fn [headers]
+                                                            (merge @response-headers headers))))
+                            request (-> request
+                                        (utils/mark-request-time :closed))]
+                        (rlog/log (assoc (rlog/request->context request)
+                                         :status (:status response)
+                                         :termination-state :instance-reservation-error))
+                        response))))))
             (catch Exception e
-              (let [{:keys [descriptor source]} (ex-data e)
-                    exception (if (= source :pr/process) (.getCause e) e)]
-                (-> (process-exception-fn track-process-error-metrics request descriptor exception)
-                    (update :headers (fn [headers]
-                                       (merge @response-headers headers))))))))))))
+              (let [{:keys [descriptor]} (ex-data e)
+                    response (-> (process-exception-fn track-process-error-metrics request descriptor e)
+                                 (update :headers (fn [headers]
+                                                    (merge @response-headers headers))))
+                    request (-> request
+                                (utils/mark-request-time :closed))]
+                (rlog/log (assoc (rlog/request->context request)
+                                 :status (:status response)
+                                 :termination-state :service-discovery-error))
+                response))))))))
```

Comment:
Seems like the same code appears 3 times, with maybe the only variant being the `:termination-state` -- should we extract a function for this?

Reply:
Done